mitodl/open-discussions
# Store channel memberships (subscriber, moderator, contributor) in django

#### Pre-Flight checklist

#### What are the relevant tickets?
Closes #1451
Closes #1453
Partially addresses #1420:
- [x] model for storing user/channel subscription status
- [x] update the subscribe API in discussions to store the subscription state of the user
- [x] whenever the subscription state is updated, update the ES profile index
- [x] update the backpopulate_default_channel management command to also store subscriptions (automatic).

#### What's this PR do?
- Adds a `ChannelGroupRole` model that will store the `Group` associated with each role for a channel.
- Adds a `ChannelSubscription` model that will store a user's subscription to a channel.
- Adds an API function `sync_channel_user_role_model` that is called whenever a channel moderator, or contributor is added or removed.
- Adds an API function `sync_channel_user_subscription_model` that is called whenever a channel subscriber is added or removed.
- Calls `search.tasks.update_author` to update the ES profile doc whenever a user's subscriber, moderator, or contributor status is changed.
- Adds a management command `backpopulate_subscriptions_roles` to iterate through all channels/users and create `ChannelGroupRole` and `ChannelSubscription` objects for them.
- Adds a `profiles.api.get_channels` function to return all channel names that a user is either moderator, contributor, or subscriber on.
- Uses the profiles.api.get_channels` function to populate `author_channel_membership` in the ES index.

#### How should this be manually tested?
- Run `docker-compose run web python manage.py migrate`
- Run `docker-compose run web python manage.py backpopulate_subscriptions_roles`
- Confirm that there is a `ChannelSubscription` object for every channel where the user is a subscriber.
- Confirm that users belong to the group associated with their role (moderators, contributors) for a channel, via `ChannelGroupRole`.
- Add and remove users as contributors or moderators for a channel.  Verify that the groups the user belongs to is updated after each change, and the profile ES doc's `author_channel_membership` field is updated.
- Run `docker-compose run web python manage.py recreate_index` and verify that `author_channel_membership` is populated correctly for profile docs.
